### PR TITLE
fix(model): fix bugs of batch generation & support min_new_tokens for inference

### PR DESCRIPTION
**Pull request by x54-729**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification
1. 修复 batch generation 时 `dones` 张量在 logic or 操作后形状变为 `(batch_size, batch_size)` 的 bug
2. 修复 batch generation 计算 rope 前 q k padding 位置移动错误的 bug
3. 修复 batch generation 计算 rope 时 q k 没有正确 pack 的 bug
4. 更新 `min_new_tokens` 功能

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/apis/inference.py [hunk 1]
```diff
@@ -461,8 +484,12 @@ def _no_beam_search_generate(
     length_penalty=1.0,
     pad_token_id=0,
```
Comment (gaoyang07) [inline internlm/apis/inference.py]: 这里的默认值也改下

File: internlm/model/modules/mha.py [hunk 2]
```diff
@@ -271,8 +271,12 @@ def _inference(self, x, inference_params, **kwargs):  # pylint: disable=W0613
                     empties = attention_mask[..., -1].sum(dim=-1)
                     indexes4q = sequence_len_offset * torch.ones(q.size(0), dtype=torch.int, device=q.device) - empties
                     indexes4k = sequence_len_offset * torch.ones(k.size(0), dtype=torch.int, device=k.device) - empties
+                    q = rearrange(q, "b s h d -> s b h d", d=self.head_dim)  # pack input
```
Comment (gaoyang07) [inline internlm/model/modules/mha.py:276]: 这里多加些注释，说明对q k做pack的目的是为了在batch维度统一做apply_rotary_emb，节省计算

Comment (gaoyang07) [reply internlm/model/modules/mha.py:276]: 下面GQA也补充类似逻辑

**Review by gaoyang07 — APPROVED** [review]
